packages/expo-video/ios/Advertisement/VideoAdsManagerIMA.swift

```diff
@@ -0,0 +1,207 @@
+// Copyright 2024-present 650 Industries. All rights reserved.
+
+// Only compile file if GoogleInteractiveMediaAds is available
+#if canImport(GoogleInteractiveMediaAds)
+import Foundation
+import GoogleInteractiveMediaAds
+
+class VideoAdsManagerIMA: NSObject, IMAAdsLoaderDelegate, IMAAdsManagerDelegate, VideoAdsManager {
+    private var adsLoader: IMAAdsLoader = IMAAdsLoader(settings: nil)
+    var adsManager: IMAAdsManager?
+    var adDisplayContainer: AdDisplayContainer?
+    var contentPlayhead: IMAAVPlayerContentPlayhead?
+    weak var delegate: VideoAdsManagerDelegate?
+    var fullscreenController: UIViewController?
+    
+    // Tracked values
+    var hasMoreAds: Bool = false
+    var isContentCompleteCalled = true
+    
+    var isPlayingAd: Bool = false {
+        didSet {
+            // Ensures the content and Ad player are syncronized
+            if isPlayingAd { player?.ref.pause() }
+            else { player?.ref.play() }
+        }
+    }
+    
+    // Keeps track of if the AVPlayer is in fullscreen
+    var isContentFullscreen: Bool = false {
+      
+      // Set the Ad view to fullscreen if content is playing
+      didSet {
+        if isPlayingAd && isContentFullscreen {
+          enterFullscreen()
+        }
+        else {
+          exitFullscreen()
+        }
+      }
+    }
+    
+    weak var player: VideoPlayer? {
+        didSet {
+            if let avPlayer = player?.ref {
+                contentPlayhead = IMAAVPlayerContentPlayhead(avPlayer: avPlayer)
+            } else {
+                contentPlayhead = nil
+            }
+        }
+    }
+    
+    func reset() {
+        hasMoreAds = false
+        isContentCompleteCalled = false
+        isPlayingAd = false
+    }
+    
+    override init() {
+        super.init()
+        self.adsLoader.delegate = self
+    }
+    
+    // This must be called before deinit due to adsManager exception - EXC_BAD_ACCESS
+    func cleanup() {
+        // Prevent Ad from playing sound during deinit
+        adsManager?.volume = 0
+        adsManager?.pause()
+        player = nil
+
+        // NOTE: Needs to run on main thread to avoid crash on iOS 17 release builds
+        DispatchQueue.main.async {
+            self.adsManager?.destroy()
+        }
+    }
+    
+    private func getTopViewController() -> UIViewController? {
+      guard let keyWindow = UIApplication.shared.connectedScenes
+        .compactMap({ ($0 as? UIWindowScene)?.windows.first(where: { $0.isKeyWindow }) })
+        .first else { return nil }
+
+      var topController = keyWindow.rootViewController
+      while let presentedController = topController?.presentedViewController {
+        topController = presentedController
+      }
+      return topController
+    }
+    
+    // Sets the ad overlay to fullscreen mode (on top of content video fullscreen)
+    public func enterFullscreen() {
```

Comment:
With the extensions above, this would now turn into e.g.:

```swift
public func enterFullscreen() {
    guard fullscreenController == nil,
          let adDisplayContainer = self.adDisplayContainer,
          let adViewController = adDisplayContainer.adContainerViewController else {
        return
    }

    fullscreenController = UIViewController.topViewController
    fullscreenController?.modalPresentationStyle = .fullScreen
    fullscreenController?.present(adViewController, animated: false)
}
```

Also, you shouldn't have to supply `completion: nil`